service/storage_service.cc

```diff
@@ -6038,6 +6077,187 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+
+    if (params.cluster_name != _db.local().get_config().cluster_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Cluster name check failed. This node cannot join the cluster "
+                                "because it expected cluster name \"{}\" and not \"{}\"",
+                                params.cluster_name,
+                                _db.local().get_config().cluster_name()),
+        };
+        co_return result;
+    }
+
+    if (params.snitch_name != _db.local().get_snitch_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Snitch name check failed. This node cannot join the cluster "
+                                "because it uses \"{}\" and not \"{}\"",
+                                params.snitch_name,
+                                _db.local().get_snitch_name()),
+        };
+        co_return result;
+    }
+
+    while (true) {
+        auto guard = co_await _group0->client().start_operation(&_abort_source);
+
+        if (_topology_state_machine._topology.is_empty()) {
+            // We need to wait until the discovery leader (us?) puts its join
+            // request. Tell the node to try again later.
+            result.result = join_node_request_result::in_progress {};
+            co_return result;
+        }
+
+        if (_topology_state_machine._topology.left_nodes.contains(params.host_id)) {
+            result.result = join_node_request_result::rejected{
+                .reason = "The node has already been removed from the cluster",
+            };
+            co_return result;
+        }
+
+        if (params.replaced_id.has_value()) {
+            auto it = _topology_state_machine._topology.normal_nodes.find(*params.replaced_id);
+            if (it == _topology_state_machine._topology.normal_nodes.end()) {
+                result.result = join_node_request_result::rejected {
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", *params.replaced_id),
+                };
+                co_return result;
+            }
+        }
+
+        // The node was not yet removed from the cluster
+
+        if (auto *p = _topology_state_machine._topology.find(params.host_id)) {
+            auto& rs = p->second;
+            if (!rs.join_cookie) {
+                // Return ok, for idempotency
+                result.result = join_node_request_result::ok {};
+                slogger.info("raft topology: join: returning ok for idempotency ({})", params.host_id);
+                co_return result;
+            } else {
+                assert(rs.state == node_state::none);
+                if (params.join_cookie != *rs.join_cookie) {
+                    // The topology coordinator still has to process the old request
+                    // and the node should reject it, only then the new request
+                    // can be placed
```

Comment:
I do not understand what old request and why do we need all this dense with cookies and retries. What exactly the scenario that you handles here? Why would a not send more then one join request? Because it restarts? Why is it not enough to check that the parameters for the new request are exactly the same as the previous one and return an error otherwise? There should be only one request for a given host_id.